scripts/operations/configuration/python_lib/api_requests.py

```diff
@@ -0,0 +1,84 @@
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+"""Shared Python function library: API requests"""
+
+import logging
+import traceback
+from typing import Callable
+
+from . import common
+
+def log_error_raise_exception(msg: str, parent_exception: Exception = None) -> None:
+    if parent_exception != None:
+        logging.debug(traceback.format_exc())
+    logging.error(msg)
+    if parent_exception == None:
+        raise common.ScriptException(msg)
+    else:        
+        raise common.ScriptException(msg) from parent_exception
+
+# Return type will depend on the particulars of the call being made
+def make_api_request_with_retries(request_method: Callable, url: str, expected_status_code: int, **request_kwargs):
+    """
+    Makes request with specified method to specified URL with specified keyword arguments (if any).
+    If the expected status code is returned, then the response body is returned (or None, if empty).
+    If a 5xx status code is returned, the request will be retried after a brief wait, a limited number
+    of times.
+    If the expected status code is never returned (or an unexpected and non-5xx status code is
+    ever returned), then raise an exception
+    """
+    try:
+        method_name = request_method.__name__.upper()
+    except Exception as e:
+        log_error_raise_exception("Unexpected error determining API request method name", e)
+
+    count=0
+    max_attempts=6
+    while count < max_attempts:
+        count+=1
+        logging.info(f"Making {method_name} request to {url}")
+        try:
+            resp = request_method(url, **request_kwargs)
+        except Exception as e:
+            log_error_raise_exception(f"Error making {method_name} request to {url}", e)
+        logging.debug(f"Response status code = {resp.status_code}")
+        if resp.status_code == expected_status_code:
+            if resp.text:            
+                try:
+                    return resp.json()
+                except Exception as e:
+                    logging.debug(f"Response reason = {resp.reason}")
+                    logging.debug(f"Response text = {resp.text}")
+                    log_error_raise_exception(f"Error parsing {url} {method_name} response as JSON", e)
+            else:
+                return None
+        logging.debug(f"Response reason = {resp.reason}")
+        logging.debug(f"Response text = {resp.text}")
+        if not 500 <= resp.status_code <= 599:
+            log_error_raise_exception("Unexpect status code received in response to API request")
```

Comment:
```suggestion
            log_error_raise_exception("Unexpected status code received in response to API request")
```

Reply:
You also could log the value of `resp.status_code` in the error message, that could be helpful.

Overall I think this code is a little brittle, for example if the API returned a 202 response we would raise an exception because it was not the expected 200? I think this is just being copied from elsewhere so I don't want to blow up this PR over this comment.

Reply:
I agree with you on including the status code in the message. I'll make that change.

I do think we would want to raise an exception if the API gave us an unexpected status code, even if it's just a different 2XX code. Or anyway, we'd want to do that on internal systems, since it would mean an API is not giving us the status code that it advertises, and so we'd want to be aware of that and make sure it isn't going to cause us problems. But I can see the argument that at least on customer systems, if we get a mismatched but still plausibly-successful status code, then we could allow it. In truth, though, that doesn't seem like logic that should be inside this function. I would think the caller would need to decide if, for their particular call, variance would be acceptable.

If this function was going to be more widely used, I guess what I would really want to do is allow the user to specify a set of status codes as acceptable, rather than only a single one. That way the caller could decide how flexible they wanted this to be. For the purposes of this PR, though, I'll just include the status code in the exception. I've opened [CASMINST-5605](https://jira-pro.its.hpecorp.net:8443/browse/CASMINST-5605) to contemplate the other sort of revisions being discussed here.

Reply:
Sure, that's good with me. I'd probably suggest an unexpected 2XX return code should be a warning or something. I just find it a little odd that we would treat e.g. 400 the same as 202.